[Bug]: Custom kernel breaks nuget

### NuGet Product Used

dotnet.exe

### Product Version

.NET SDK Version: 6.0.302 Commit: c857713418

### Worked before?

.NET SDK Version: 6.0.302 Commit: c857713418

### Impact

It's more difficult to complete my work

### Repro Steps & Context

It would be nice to be able to disable the sending of unnecessary information through the UserAgent.
```cs
...
///For Unix this currently comes from "uname -srv"
...
    RuntimeInformation.OSDescription
```
-- from dotnet/sdk/blob/main/src/Cli/dotnet/Telemetry/TelemetryCommonProperties.cs

In addition to what is shown in the "verbose logs", performing `dotnet restore` in a project that already has a `PackageReference` in .csproj gives an error:
`error NU1301: Unable to load the service index for source https://api.nuget.org/v3/index.json.`

### Verbose Logs

```shell
$ ls
$ dotnet new console
$ dotnet add package Newtonsoft.Json
  Determining projects to restore...
  Writing /tmp/tmpUqFyzk.tmp
info : Adding PackageReference for package 'Newtonsoft.Json' into project '/home/Killfrra/Documents/test/test.csproj'.
error: Unable to load the service index for source https://api.nuget.org/v3/index.json.
error:   The format of value '(Linux 5.18.0-14.2-liquorix-amd64 #1 ZEN SMP PREEMPT_DYNAMIC liquorix 5.18-16ubuntu1~jammy (2022-0)' is invalid.


Usage: NuGet.CommandLine.XPlat.dll package add [options]

Options:
  -h|--help               Show help information
  --force-english-output  Forces the application to run using an invariant, English-based culture.
  --package               Id of the package to be added.
  --version               Version of the package to be added.
  -d|--dg-file            Path to the dependency graph file to be used to restore preview and compatibility check.
  -p|--project            Path to the project file.
  -f|--framework          Frameworks for which the package reference should be added.
  -n|--no-restore         Do not perform restore preview and compatibility check. The added package reference will be unconditional.
  -s|--source             Specifies NuGet package sources to use during the restore.
  --package-directory     Directory to restore packages in.
  --interactive           Allow the command to block and require manual action for operations like authentication.
  --prerelease            Allows prerelease packages to be installed.

$ uname -srv
Linux 5.18.0-14.2-liquorix-amd64 #1 ZEN SMP PREEMPT_DYNAMIC liquorix 5.18-16ubuntu1~jammy (2022-0
$
```

--- a/src/NuGet.Core/NuGet.Protocol/UserAgentStringBuilder.cs
+++ b/src/NuGet.Core/NuGet.Protocol/UserAgentStringBuilder.cs
@@ -36,7 +36,12 @@ namespace NuGet.Protocol.Core.Types
 
         public UserAgentStringBuilder WithOSDescription(string osInfo)
         {
-            _osInfo = osInfo;
+#if NETCOREAPP2_0_OR_GREATER
+            _osInfo = osInfo.Replace("(", @"\(", StringComparison.Ordinal).Replace(")", @"\)", StringComparison.Ordinal);
+#else
+            _osInfo = osInfo.Replace("(", @"\(").Replace(")", @"\)");
+#endif
+
             return this;
         }
 

--- a/test/NuGet.Core.Tests/NuGet.Protocol.Tests/UserAgentStringBuilderTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Protocol.Tests/UserAgentStringBuilderTests.cs
@@ -103,11 +103,14 @@ namespace NuGet.Protocol.Tests
         [InlineData("Custom Kernel 123)")]
         public void Build_OsDescriptionWithUnmatchedParenthesis_IsValid(string osDescription)
         {
-            var target = new UserAgentStringBuilder();
+            // Arrange
+            UserAgentStringBuilder target = new();
 
-            var result = target.WithOSDescription(osDescription).Build();
+            // Act
+            string result = target.WithOSDescription(osDescription).Build();
 
-            var httpRequest = new HttpRequestMessage();
+            // Assert
+            HttpRequestMessage httpRequest = new();
             httpRequest.Headers.Add("User-Agent", result);
         }
     }

--- a/test/NuGet.Core.Tests/NuGet.Protocol.Tests/UserAgentStringBuilderTests.cs
+++ b/test/NuGet.Core.Tests/NuGet.Protocol.Tests/UserAgentStringBuilderTests.cs
@@ -1,6 +1,7 @@
 // Copyright (c) .NET Foundation. All rights reserved.
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System.Net.Http;
 using NuGet.Protocol.Core.Types;
 using Xunit;
 using Xunit.Abstractions;
@@ -95,6 +96,19 @@ namespace NuGet.Protocol.Tests
             Assert.True(userAgentString2.Contains(builder.NuGetClientVersion));
             Assert.True(userAgentString3.Contains(builder.NuGetClientVersion));
             Assert.True(userAgentString4.Contains(builder.NuGetClientVersion));
+        }
+
+        [Theory]
+        [InlineData("Custom Kernel (123")]
+        [InlineData("Custom Kernel 123)")]
+        public void Build_OsDescriptionWithUnmatchedParenthesis_IsValid(string osDescription)
+        {
+            var target = new UserAgentStringBuilder();
+
+            var result = target.WithOSDescription(osDescription).Build();
+
+            var httpRequest = new HttpRequestMessage();
+            httpRequest.Headers.Add("User-Agent", result);
         }
     }
 }

